spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,491 @@
+# -*- coding: utf-8 -*-
+# %% [markdown]
+# # ✂️ _Spam_ — Data Slicing Tutorial
+#
+# In real-world applications, some model outcomes are often more important than others — e.g. vulnerable cyclist detections in an autonomous driving task, or, in our running **spam** application, potentially malicious link redirects to external websites.
+#
+# Traditional machine learning systems optimize for overall quality, which may be too coarse-grained: models that achieve high overall performance might produce unacceptable failure rates on critical slices of the data — data subsets that might correspond to vulnerable cyclist detection in an autonomous driving task, or in our running spam detection application, external links to potentially malicious websites.
+#
+# In this tutorial, we introduce _Slicing Functions (SFs)_ as a programming interface to:
+# 1. **Monitor** application-critical data slices
+# 2. **Improve model performance** on slices
+
+# %% [markdown]
+# First, we'll set up our notebook for reproducibility and proper logging.
+
+# %%
+import logging
+import os
+import pandas as pd
+from snorkel.utils import set_seed
+
+# For reproducibility
+os.environ["PYTHONHASHSEED"] = "0"
+set_seed(111)
+
+# Make sure we're running from the spam/ directory
+if os.path.basename(os.getcwd()) == "snorkel-tutorials":
+    os.chdir("spam")
+
+# To visualize logs
+logger = logging.getLogger()
+logger.setLevel(logging.WARNING)
+
+# Show full columns for viewing data
+pd.set_option("display.max_colwidth", -1)
+
+# %% [markdown]
+# _Note:_ this tutorial differs from the labeling tutorial in that we use ground truth labels in the train split for demo purposes.
+# In practice, data slicing is agnostic to the _training labels_ used as inputs — you can use Snorkel-generated labels as inputs to this pipeline.
+
+# %%
+from utils import load_spam_dataset
+
+df_train, df_valid, df_test = load_spam_dataset(
+    load_train_labels=True, include_dev=False
+)
+
+
+# %% [markdown]
+# ## 1. Train a discriminative model
+#
+# To start, we'll initialize a discriminative model using our [`SnorkelClassifier`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.html).
```

Comment:
links are incorrect. should be https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel.classification.MultitaskClassifier. going to skip these comments for the rest.